pkgs/misc/drivers/xow/default.nix

```diff
@@ -20,7 +25,14 @@ stdenv.mkDerivation rec {
     "MODPDIR=${placeholder ''out''}/lib/modprobe.d"
     "SYSDDIR=${placeholder ''out''}/lib/systemd/system"
```

Comment:
Looks like all these paths changed:

https://github.com/medusalix/xow/commit/29999d577a1a775a028ae9981c5cc37d86e7cec1

Not sure if it is correct.

Reply:
Yeah, my suspicion was right – the previous values were correct. Reading `systemd.unit (5)`, the path `/etc/systemd/system` is described as “System units created by the administrator”, whereas the previous value `/lib/systemd/system` had “System units installed by the distribution package manager”.

Reply:
And also https://unix.stackexchange.com/questions/206315/whats-the-difference-between-usr-lib-systemd-system-and-etc-systemd-system. `/etc` is reserved for sysadmins, packages should install to `/lib`

Reply:
When you install xow from source, there is usually no package manager involved, so the installation to `/etc` is in fact correct. Overwriting files placed into `/lib` by a package manager would be confusing.

Reply:
Then it should IMHO go to `/usr/local/lib/systemd/system` – “System units installed by the administrator” resp. `$(PREFIX)/lib/systemd/system` since administrator does not create them.

Reply:
udev has the same structure ([Debian](http://manpages.ubuntu.com/manpages/groovy/man7/udev.7.html), [non-Debian](https://www.man7.org/linux/man-pages/man7/udev.7.html)). Maybe we should have `rootprefix` like systemd does: https://github.com/systemd/systemd/blob/b7d81d19cc13987319d61782d3b4c7bff8b0b4da/meson.build#L79-L86

```make
PREFIX := /usr/local
# Debian distros use /lib instead of /usr/lib
ROOTPREFIX := $(PREFIX)
BINDIR := $(PREFIX)/bin
UDEVDIR := $(ROOTPREFIX)/lib/udev/rules.d
MODLDIR := $(ROOTPREFIX)/lib/modules-load.d
MODPDIR := $(ROOTPREFIX)/lib/modprobe.d
SYSDDIR := $(ROOTPREFIX)/lib/systemd/system
```

Then normal distros would set `PREFIX=/usr`, NixOS would set `PREFIX=${placeholder "out"}` and Debian would set `PREFIX=/usr ROOTPREFIX=/`.

Reply:
According to https://www.freedesktop.org/software/systemd/man/modules-load.d.html, `modules-load.d` supports that too:

> Packages should install their configuration files in `/usr/lib/` (distribution packages) or `/usr/local/lib/` (local installs). Files in `/etc/` are reserved for the local administrator, who may use this logic to override the configuration files installed by vendor packages. It is recommended to prefix all filenames with a two-digit number and a dash, to simplify the ordering of the files.

So the only odd one out is [`modprobe.d`](https://man7.org/linux/man-pages/man5/modprobe.d.5.html). But there it might still be preferable to override the `/lib/xow-blacklist.conf` file installed by package manager (since package manager can repair that). If user administrator sets up their own blacklist in `/etc`, overriding it would not be so painless.

Reply:
> Then it should IMHO go to /usr/local/lib/systemd/system – “System units installed by the administrator” resp. $(PREFIX)/lib/systemd/system since administrator does not create them.

Yeah, I fully agree with that, but there's one problem: If I make changes to these files and they get installed to `/usr/local/*`, the old configs in `/etc/*` will still override all of the newer files. It's basically a breaking change, as users have to make sure they delete the old files manually.

Reply:
Can that not be fixed by mentioning it in release news? Hopefully, most people would use distro package rather than modifying their system manually.

Reply:
Most people build from source, as there's no Debian (and likewise Ubuntu) package available. People don't even read the README properly, so I'm not particularly keen on providing technical support for those whose installations I just broke. I get enough of these issue reports already.
I hope this explains my reasoning.

Reply:
I agree with what @jtojnar is suggesting. I hope that moving the install location isn't some one-way street.
I think you can make a pinned issue warning everyone to remove previous files. I'd be happy to help filter out the end user problems that come from this if it means fixing it.

Reply:
Yes, I also gave this a second thought and I stumbled upon [this](https://git.kernel.org/pub/scm/utils/kernel/module-init-tools/module-init-tools.git/commit/?id=9454d710137be3799f343cc9d0f833f0802e2111) commit where additional config directories were added to `modprobe`. That means, I would be able to implement the solution @jtojnar suggested in an even more straight forward fashion.
I apologize for my slightly exaggerated description of the issues that would arise with such a change. It certainly won't be as bad as I initially thought, given that I don't plan on making changes to the config files very often.

@wgpierce I truly appreciate your support and your suggestion. If I need to add or remove something to the config files in the future, I'll definitely make sure to create a pinned issue about it.

Reply:
It might also be useful to have a `make uninstall` target that uninstalls all files from all previous versions if that seems worth the effort.

Reply:
I think it might be confusing if `make uninstall` removes files that might have not necessarily been installed by xow. Maybe an additional target like `purge`?

@jtojnar Do you think moving the xow executable from `/usr/local/bin` to `/usr/local/libexec` might also be preferable? The recommended way to launch xow is via the systemd service and running xow manually (in addition to the service) usually causes problems.

Reply:
> I think it might be confusing if make uninstall removes files that might have not necessarily been installed by xow (maybe the system administrator modified them).

Yes, that is why there is a distinction between `/etc` and the rest of the file system. And also why users are discouraged from running `make install` and instead are supposed to create packages for software they want to use. (At least that was the case on Arch when I used it, and NixOS does not even allow installing stuff to `/usr`.)

But with 0.5 the seals were broken and we can no longer be sure who is the owner of files in `/etc` (unless we decide to rename the files to something like `xow-6`).

> Do you think moving the xow executable from /usr/local/bin to /usr/local/libexec might also be preferable?

Yeah, moving it to `libexecdir` is a good idea. It is what most daemons do and what I would expect from https://www.gnu.org/prep/standards/html_node/Directory-Variables.html#index-libexecdir

Reply:
`make purge` makes sense to me if you want to catch these cases. Note that any other file moves like https://github.com/medusalix/xow/commit/9e86c529afab1ffde40e2554e5eee1ecfcbf8c7a are subject to leaving behind artifacts as well.

Reply:
> Yeah, moving it to libexecdir is a good idea.

I wonder why I didn't notice that the first time I read the docs. It makes things *so much* easier.
I recommend people to run `make uninstall` for every update they do (before installing the new version), so if they stick to that it will cause absolutely no problems for them. Users that didn't read the README might wonder why they're no longer running the newest version even after running `make install`, so I hope they'll then see the pinned issue.

Reply:
```make
PREFIX := /usr/local
LIBEXECDIR := $(PREFIX)/libexec
LIBDIR := $(PREFIX)/lib
UDEVDIR := $(LIBDIR)/udev/rules.d
MODLDIR := $(LIBDIR)/modules-load.d
MODPDIR := $(LIBDIR)/modprobe.d
SYSDDIR := $(LIBDIR)/systemd/system
```

@jtojnar This is what I've come up with regarding my last comment. I don't want to make changes to the installation directory a third time, so I need your full approval before I commit this.

Reply:
Looks like the rules files aren't getting loaded correctly at the moment, mt76x2u grabs the dongle before xow can. Do the paths to them need to be specified in some option in the nixos module file? It was working before though, and I don't see any relevant-looking option in the [browser](https://nixos.org/nixos/options.html).

Reply:
@medusalix That looks good, except I am not sure about the modprobe. For example, `kmod` does not seem to support `/usr/local`: https://git.kernel.org/pub/scm/utils/kernel/kmod/kmod.git/tree/libkmod/libkmod.c?id=f5434cf5fc5b567359e1b9207bbab24c6782cfbd#n65 It would be best to test if it works there first.

Reply:
@jtojnar I found [this](https://salsa.debian.org/md/kmod/-/commit/1a56e3b10f493b93ff09d3f3ec5e3ac73cd13115) commit on Debian's `kmod` packaging repository. It mentions `systemd-modules-load`, which seems to include `/usr/local/lib/modules-load.d` in its search path (see [this](https://github.com/systemd/systemd/commit/223a35581604897185c46dfc184b0fa32916d59b) commit).